pkg/generators/interface.go

```diff
@@ -0,0 +1,14 @@
+package generators
+
+import (
+	argoprojiov1alpha1 "github.com/argoproj-labs/applicationset/api/v1alpha1"
+	argov1alpha1 "github.com/argoproj/argo-cd/pkg/apis/application/v1alpha1"
+)
+
+// Generator defines the interface implemented by all ApplicationSet generators.
+type Generator interface {
+	// GenerateApplications interprets the ApplicationSet and generates all relevant Applications.
+	// The expected / desired list of Applications is returned, it then needs to be reconciled
+	// against the current state of the Applications in the cluster.
+	GenerateApplications(appSet *argoprojiov1alpha1.ApplicationSet) ([]argov1alpha1.Application, error)
+}
```

Comment:
I thought it might be helpful to break this into an interface and implement for each type of generator, to help us work without causing git conflicts as much as possible.

Returns the Applications (desired state) to help with unit testing, something else will need to reconcile these with the actual state of the Applications currently.

Common code can be shared amongst generators using functions in this package, but my suspicion is the implementations of each generator will be unique enough to isolate them as separate implementations.